public/app/features/trails/utils.ts

```diff
@@ -115,3 +118,42 @@ export function getFilters(scene: SceneObject) {
   }
   return null;
 }
+
+// frontend hardening limit
+const MAX_ADHOC_VARIABLE_OPTIONS = 10000;
+
+export function limitAdhocProviders(
+  filtersVariable: SceneVariable<SceneVariableState> | null,
+  datasourceHelper: MetricDatasourceHelper
+) {
+  if (filtersVariable instanceof AdHocFiltersVariable) {
+    filtersVariable.setState({
+      getTagKeysProvider: async (
+        variable: AdHocFiltersVariable,
+        currentKey: string | null
+      ): Promise<{
+        replace?: boolean;
+        values: GetTagResponse | MetricFindValue[];
```

Comment:
It would be nice if we have a type for this one so we can reuse.

Reply:
This makes sense, but I will go with the implicit typing that [@NWRichmond mentioned below](https://github.com/grafana/grafana/pull/94036#pullrequestreview-2341493935) as this is the convention in Scenes. I can not think of another case outside of this where we would reuse the type for what is returned in an adhoc filter variable.